### PR TITLE
Fix ranef/coef output

### LINKED ISSUE #370: stan_lmer with uncorrelated intercepts and slopes
#### Summary:
It looks like `stan_lmer/stan_glmer` doesn't provide the estimated random effects when using the `lme4` syntax of no correlation between intercepts slopes - `(1|group) + (0 + x|group)`. 

#### Description:
The associated methods appear to ignore the random slopes output when using `lme4`-style syntax to specify a model with random intercepts and slopes but with no correlation. Specifically, neither `ranef` nor `coef` provide varying slopes in this case.  Oddly, the summary/print output suggests the model was in fact estimated as the syntax would imply.

#### Reproducible Steps:

``` r
library(lme4);  library(rstanarm)
#> Loading required package: Matrix
#> Loading required package: Rcpp
#> Registered S3 methods overwritten by 'ggplot2':
#>   method         from 
#>   [.quosures     rlang
#>   c.quosures     rlang
#>   print.quosures rlang
#> Registered S3 method overwritten by 'xts':
#>   method     from
#>   as.zoo.xts zoo
#> rstanarm (Version 2.18.2, packaged: 2018-11-08 22:19:38 UTC)
#> - Do not expect the default priors to remain the same in future rstanarm versions.
#> Thus, R scripts should specify priors explicitly, even if they are just the defaults.
#> - For execution on a local, multicore CPU with excess RAM we recommend calling
#> options(mc.cores = parallel::detectCores())
#> - Plotting theme set to bayesplot::theme_default().

f1 = Reaction ~ Days + (1 | Subject)  + (0 + Days | Subject)
f2 = Reaction ~ Days + (Days || Subject)

model0 = lmer(f1, sleepstudy)

model1 = stan_lmer(f1, data = sleepstudy, iter = 1000, chains = 1)
#> 
#> SAMPLING FOR MODEL 'continuous' NOW (CHAIN 1).
#> Chain 1: 
#> Chain 1: Gradient evaluation took 0 seconds
#> Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 0 seconds.
#> Chain 1: Adjust your expectations accordingly!
#> Chain 1: 
#> Chain 1: 
#> Chain 1: Iteration:   1 / 1000 [  0%]  (Warmup)
#> Chain 1: Iteration: 100 / 1000 [ 10%]  (Warmup)
#> Chain 1: Iteration: 200 / 1000 [ 20%]  (Warmup)
#> Chain 1: Iteration: 300 / 1000 [ 30%]  (Warmup)
#> Chain 1: Iteration: 400 / 1000 [ 40%]  (Warmup)
#> Chain 1: Iteration: 500 / 1000 [ 50%]  (Warmup)
#> Chain 1: Iteration: 501 / 1000 [ 50%]  (Sampling)
#> Chain 1: Iteration: 600 / 1000 [ 60%]  (Sampling)
#> Chain 1: Iteration: 700 / 1000 [ 70%]  (Sampling)
#> Chain 1: Iteration: 800 / 1000 [ 80%]  (Sampling)
#> Chain 1: Iteration: 900 / 1000 [ 90%]  (Sampling)
#> Chain 1: Iteration: 1000 / 1000 [100%]  (Sampling)
#> Chain 1: 
#> Chain 1:  Elapsed Time: 2.682 seconds (Warm-up)
#> Chain 1:                0.553 seconds (Sampling)
#> Chain 1:                3.235 seconds (Total)
#> Chain 1:

model2 = stan_lmer(f2, data = sleepstudy, iter = 1000, chains = 1)
#> 
#> SAMPLING FOR MODEL 'continuous' NOW (CHAIN 1).
#> Chain 1: 
#> Chain 1: Gradient evaluation took 0 seconds
#> Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 0 seconds.
#> Chain 1: Adjust your expectations accordingly!
#> Chain 1: 
#> Chain 1: 
#> Chain 1: Iteration:   1 / 1000 [  0%]  (Warmup)
#> Chain 1: Iteration: 100 / 1000 [ 10%]  (Warmup)
#> Chain 1: Iteration: 200 / 1000 [ 20%]  (Warmup)
#> Chain 1: Iteration: 300 / 1000 [ 30%]  (Warmup)
#> Chain 1: Iteration: 400 / 1000 [ 40%]  (Warmup)
#> Chain 1: Iteration: 500 / 1000 [ 50%]  (Warmup)
#> Chain 1: Iteration: 501 / 1000 [ 50%]  (Sampling)
#> Chain 1: Iteration: 600 / 1000 [ 60%]  (Sampling)
#> Chain 1: Iteration: 700 / 1000 [ 70%]  (Sampling)
#> Chain 1: Iteration: 800 / 1000 [ 80%]  (Sampling)
#> Chain 1: Iteration: 900 / 1000 [ 90%]  (Sampling)
#> Chain 1: Iteration: 1000 / 1000 [100%]  (Sampling)
#> Chain 1: 
#> Chain 1:  Elapsed Time: 3.209 seconds (Warm-up)
#> Chain 1:                0.668 seconds (Sampling)
#> Chain 1:                3.877 seconds (Total)
#> Chain 1:

ranef(model0)
#> $Subject
#>     (Intercept)        Days
#> 308   1.5116973   9.3237308
#> 309 -40.3720066  -8.5995358
#> 310 -39.1795117  -5.3880715
#> 330  24.5188072  -4.9686809
#> 331  22.9141930  -3.1939310
#> 332   9.2217762  -0.3084673
#> 333  17.1557243  -0.2871512
#> 334  -7.4516633   1.1159907
#> 335   0.5798353 -10.9062401
#> 337  34.7661741   8.6279628
#> 349 -25.7538155   1.2806250
#> 350 -13.8653871   6.7565202
#> 351   4.9161750  -3.0751926
#> 352  20.9281601   3.5123758
#> 369   3.2584763   0.8730848
#> 370 -26.4756822   4.9838147
#> 371   0.9057286  -1.0053150
#> 372  12.4213189   1.2584805
#> 
#> with conditional variances for "Subject"

ranef(model1)
#> $Subject
#>     (Intercept)
#> 308   0.4969372
#> 309 -40.2864081
#> 310 -40.2821158
#> 330  23.4173588
#> 331  21.4068756
#> 332   8.0675373
#> 333  15.9576733
#> 334  -7.8245233
#> 335  -0.1057223
#> 337  32.4620070
#> 349 -26.1875155
#> 350 -13.6666587
#> 351   3.9922127
#> 352  19.0864617
#> 369   1.7823361
#> 370 -26.7819893
#> 371  -0.4712585
#> 372  10.6913008

ranef(model2)
#> $Subject
#>     (Intercept)
#> 308   2.7515793
#> 309 -38.8264329
#> 310 -37.5865994
#> 330  24.1302539
#> 331  22.6735069
#> 332   9.0259098
#> 333  15.3172050
#> 334  -8.0149340
#> 335   0.1518824
#> 337  34.2369830
#> 349 -25.4775255
#> 350 -13.5690587
#> 351   5.4329586
#> 352  20.1272949
#> 369   2.9226653
#> 370 -25.5256030
#> 371  -0.6972952
#> 372  11.6940454

coef(model0)
#> $Subject
#>     (Intercept)       Days
#> 308    252.9168 19.7910167
#> 309    211.0331  1.8677501
#> 310    212.2256  5.0792144
#> 330    275.9239  5.4986051
#> 331    274.3193  7.2733550
#> 332    260.6269 10.1588187
#> 333    268.5608 10.1801347
#> 334    243.9534 11.5832767
#> 335    251.9849 -0.4389541
#> 337    286.1713 19.0952488
#> 349    225.6513 11.7479110
#> 350    237.5397 17.2238062
#> 351    256.3213  7.3920934
#> 352    272.3333 13.9796618
#> 369    254.6636 11.3403708
#> 370    224.9294 15.4511007
#> 371    252.3108  9.4619709
#> 372    263.8264 11.7257665
#> 
#> attr(,"class")
#> [1] "coef.mer"

coef(model1)
#> $Subject
#>     (Intercept)     Days
#> 308    252.7841 10.35827
#> 309    212.0008 10.35827
#> 310    212.0051 10.35827
#> 330    275.7045 10.35827
#> 331    273.6940 10.35827
#> 332    260.3547 10.35827
#> 333    268.2448 10.35827
#> 334    244.4627 10.35827
#> 335    252.1815 10.35827
#> 337    284.7492 10.35827
#> 349    226.0997 10.35827
#> 350    238.6205 10.35827
#> 351    256.2794 10.35827
#> 352    271.3736 10.35827
#> 369    254.0695 10.35827
#> 370    225.5052 10.35827
#> 371    251.8159 10.35827
#> 372    262.9785 10.35827
#> 
#> attr(,"class")
#> [1] "coef.mer"

coef(model2)
#> $Subject
#>     (Intercept)     Days
#> 308    254.3371 10.49802
#> 309    212.7591 10.49802
#> 310    213.9989 10.49802
#> 330    275.7158 10.49802
#> 331    274.2590 10.49802
#> 332    260.6114 10.49802
#> 333    266.9027 10.49802
#> 334    243.5706 10.49802
#> 335    251.7374 10.49802
#> 337    285.8225 10.49802
#> 349    226.1080 10.49802
#> 350    238.0165 10.49802
#> 351    257.0185 10.49802
#> 352    271.7128 10.49802
#> 369    254.5082 10.49802
#> 370    226.0599 10.49802
#> 371    250.8882 10.49802
#> 372    263.2796 10.49802
#> 
#> attr(,"class")
#> [1] "coef.mer"


model0
#> Linear mixed model fit by REML ['lmerMod']
#> Formula: Reaction ~ Days + (1 | Subject) + (0 + Days | Subject)
#>    Data: sleepstudy
#> REML criterion at convergence: 1743.669
#> Random effects:
#>  Groups    Name        Std.Dev.
#>  Subject   (Intercept) 25.050  
#>  Subject.1 Days         5.989  
#>  Residual              25.565  
#> Number of obs: 180, groups:  Subject, 18
#> Fixed Effects:
#> (Intercept)         Days  
#>      251.41        10.47

model1
#> stan_lmer
#>  family:       gaussian [identity]
#>  formula:      Reaction ~ Days + (1 | Subject) + (0 + Days | Subject)
#>  observations: 180
#> ------
#>             Median MAD_SD
#> (Intercept) 252.3    6.6 
#> Days         10.4    1.7 
#> 
#> Auxiliary parameter(s):
#>       Median MAD_SD
#> sigma 25.7    1.5  
#> 
#> Error terms:
#>  Groups   Name        Std.Dev.
#>  Subject  (Intercept) 26.5    
#>  Subject  Days         6.7    
#>  Residual             25.8    
#> Num. levels: Subject 18 
#> 
#> Sample avg. posterior predictive distribution of y:
#>          Median MAD_SD
#> mean_PPD 298.3    2.7 
#> 
#> ------
#> * For help interpreting the printed output see ?print.stanreg
#> * For info on the priors used see ?prior_summary.stanreg

model2
#> stan_lmer
#>  family:       gaussian [identity]
#>  formula:      Reaction ~ Days + ((1 | Subject) + (0 + Days | Subject))
#>  observations: 180
#> ------
#>             Median MAD_SD
#> (Intercept) 251.6    6.8 
#> Days         10.5    1.7 
#> 
#> Auxiliary parameter(s):
#>       Median MAD_SD
#> sigma 25.8    1.5  
#> 
#> Error terms:
#>  Groups   Name        Std.Dev.
#>  Subject  (Intercept) 26.1    
#>  Subject  Days         6.5    
#>  Residual             25.9    
#> Num. levels: Subject 18 
#> 
#> Sample avg. posterior predictive distribution of y:
#>          Median MAD_SD
#> mean_PPD 298.6    2.7 
#> 
#> ------
#> * For help interpreting the printed output see ?print.stanreg
#> * For info on the priors used see ?prior_summary.stanreg

## compare with
## ranef(stan_lmer(Reaction ~ Days + (Days | Subject), data = sleepstudy, iter = 500, chains = 1))
```

<sup>Created on 2019-06-19 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

#### RStanARM Version:
2.18.2

#### R Version:
3.6.0

#### Operating System:
Seen on OS Mojave 10.14.5 and Windows 7

### PR DESCRIPTION
**Pull request by jgabry**
Fixes #370 by having `ranef.stanreg()` first create a ranef template using lme4 and then replace the elements with the corresponding Stan parameter estimates.  

### Example

```r
# These are the same models from issue 370
f1 <- Reaction ~ Days + (1 | Subject)  + (0 + Days | Subject)
f2 <- Reaction ~ Days + (Days || Subject)

model0 <- lmer(f1, sleepstudy)
model1 <- stan_lmer(f1, data = sleepstudy, cores = 4)
model2 <- stan_lmer(f2, data = sleepstudy, cores = 4)
```
```r
> ranef(model0)
$Subject
    (Intercept)        Days
308   1.5116973   9.3237308
309 -40.3720066  -8.5995358
310 -39.1795117  -5.3880715
330  24.5188072  -4.9686809
331  22.9141930  -3.1939310
332   9.2217762  -0.3084673
333  17.1557243  -0.2871512
334  -7.4516633   1.1159907
335   0.5798353 -10.9062401
337  34.7661741   8.6279628
349 -25.7538155   1.2806250
350 -13.8653871   6.7565202
351   4.9161750  -3.0751926
352  20.9281601   3.5123758
369   3.2584763   0.8730848
370 -26.4756822   4.9838147
371   0.9057286  -1.0053150
372  12.4213189   1.2584805

with conditional variances for “Subject” 
> ranef(model1)
$Subject
    (Intercept)        Days
308    1.168278   9.3266554
309  -39.259358  -8.7801971
310  -37.820457  -5.4892819
330   24.135714  -4.9490920
331   22.616169  -3.2382528
332    8.908237  -0.2772714
333   17.006650  -0.3265312
334   -7.064768   1.0919659
335    1.282598 -10.9757874
337   33.862448   8.6580530
349  -24.913804   1.1811923
350  -13.728148   6.6852530
351    5.043708  -3.1333312
352   20.860706   3.5302807
369    3.256299   0.8247093
370  -25.800626   4.9535799
371    1.008259  -1.0457000
372   12.169819   1.2224220

with conditional variances for “Subject” 
> ranef(model2)
$Subject
    (Intercept)        Days
308   1.5169717   9.3588436
309 -38.9895072  -8.6444823
310 -38.1888151  -5.3082307
330  23.7753771  -4.7876148
331  21.9490893  -2.9483002
332   9.2418477  -0.2747270
333  16.7850602  -0.1925011
334  -6.9982244   1.1506574
335   1.1093192 -10.8977349
337  33.9905269   8.7487089
349 -24.9314499   1.1559632
350 -13.3293416   6.6786021
351   4.9678099  -3.0121604
352  20.4254964   3.6360025
369   3.2923601   0.8826810
370 -25.6056696   4.9622716
371   0.6943152  -0.9977069
372  12.2368671   1.3239543

with conditional variances for “Subject” 
```

### CODE REVIEW

File: R/stanreg-methods.R [hunk 1]
```diff
@@ -235,32 +235,77 @@ nsamples.stanreg <- function(object, ...) {
...
-  nb <- nc * vapply(levs, length, 1L)
-  nbseq <- rep.int(seq_along(nb), nb)
-  ml <- split(ans, nbseq)
-  for (i in seq_along(ml)) {
-    ml[[i]] <- matrix(ml[[i]], ncol = nc[i], byrow = TRUE, 
-                      dimnames = list(NULL, cnms[[i]]))
+  ans <- object$stan_summary[b_names(all_names), select_median(object$algorithm)]
+  ans <- ans[!grepl("_NEW_", names(ans), fixed = TRUE)] 
+  
+  out <- ranef_template(object)
```
Comment (bgoodri) [inline R/stanreg-methods.R]: Can we call `ranef_template` first and avoid the monstrosity that is the `b_names` function?

Comment (jgabry) [reply R/stanreg-methods.R]: Yeah I’ll change that

**Comment by jgabry** [comment]
@bgoodri Ok I think all set. Can you give it another look through before merging?

**Comment by bgoodri** [comment]
I mean: Can we use this technique to eliminate the `b_names` function entirely and just use whatever naming convention **lme4** uses for the group-specific terms?

**Comment by jgabry** [comment]
Can you clarify a bit more? Aren't we already basically using the naming conventions lme4 is using for group-specific terms? All `b_names` does is grep for parameter names starting with `b[`

https://github.com/stan-dev/rstanarm/blob/6ff0f9604fdee1f72cf56a88c24163eb4b67132f/R/misc.R#L381-L383

so I don't see how this template thing is a replacement for that. Did you mean a different function other than `b_names`?

**Comment by bgoodri** [comment]
I basically mean that if we delete the `b_names` function, what breaks and
if so, can it be fixed? Here are the places that b_names is used:

R/log_lik.R:342:    b_sel <- if (is.null(nms)) b_names(colnames(stanmat))
else nms$y_b[[m]]
R/misc.R:377:b_names <- function(x, ...) {
R/misc.R:1048:  b <- b_names(x, ...)
R/misc.R:1049:  y_b <- lapply(1:M, function(m) b_names_M(x, m, stub = stub,
...))
R/misc.R:1062:b_names_M <- function(x, submodel = NULL, stub = "Long", ...)
{
R/print-and-summary.R:584:    pars <- setdiff(pars, b_names(pars, value =
TRUE))
R/print-and-summary.R:602:  nms_tmp_b <- b_names(nms_tmp, value = TRUE)
R/print-and-summary.R:728:      pars2 <- c(pars2,
b_names(rownames(object$stan_summary), value = TRUE))
R/simulate_b_pars.R:107:  sel <- b_names(colnames(stanmat))    # stanmat
cols containing old b pars
R/simulate_b_pars.R:191:  pars$stanmat <- pars$stanmat[,
-b_names(colnames(pars$stanmat)), drop = FALSE]
R/stan_gamm4.R:234:    colnames(Z) <- b_names(names(stanfit), value = TRUE)
R/stan_glmer.R:160:  colnames(Z) <- b_names(names(stanfit), value = TRUE)
R/stan_nlmer.R:173:  colnames(Z) <- b_names(names(stanfit), value = TRUE)
R/stanmvreg-methods.R:396:  coefs <- lapply(coefs, function(x)
x[b_names(names(x), invert = TRUE)])
R/stanmvreg-methods.R:423:    sel <- b_names_M(all_names, x, stub = stub)
R/stanreg-methods.R:210:  coefs[b_names(names(coefs), invert = TRUE)]
R/stanreg-methods.R:238:  sel <- b_names(all_names)

Also, I'm not sure we name things in exactly the same way that lme4 does.
They get added by that pad_ReTrms() function.

**Comment by jgabry** [comment]
Ok I see what you mean. I will look into that. But this bugfix no longer uses `b_names` and so would be unaffected by that, so if you think it's implemented correctly I think go ahead and merge it and I'll look into `b_names` separately.

**Pull request merged**